teos/src/api/tor.rs

```diff
@@ -0,0 +1,107 @@
+use std::io::{Error, ErrorKind};
+use std::net::SocketAddr;
+use tokio::net::TcpStream;
+use tokio::time::{sleep, Duration};
+use torut::control::UnauthenticatedConn;
+use torut::onion::TorSecretKeyV3;
+use triggered::Listener;
+
+/// Expose an onion service that re-directs to the public api.
+pub async fn expose_onion_service(
+    tor_control_port: u16,
+    api_port: u16,
+    onion_port: u16,
+    shutdown_signal_tor: Listener,
+) -> Result<(), Error> {
+    let stream = connect_tor_cp(format!("127.0.0.1:{}", tor_control_port).parse().unwrap())
+        .await
+        .map_err(|e| Error::new(ErrorKind::ConnectionRefused, e))?;
+
+    let mut utc = UnauthenticatedConn::new(stream);
+
+    let info = utc
+        .load_protocol_info()
+        .await
+        .map_err(|e| Error::new(ErrorKind::ConnectionRefused, e))?;
+
+    let ad = info.make_auth_data()?.expect("failed to make auth data");
+
+    utc.authenticate(&ad).await.map_err(|_| {
+        Error::new(
+            ErrorKind::PermissionDenied,
+            "failed to authenticate with Tor",
+        )
+    })?;
+
+    let mut ac = utc.into_authenticated().await;
+
+    ac.set_async_event_handler(Some(|_| async move { Ok(()) }));
+
+    let key = TorSecretKeyV3::generate();
+
+    let _ = ac
+        .add_onion_v3(
+            &key,
+            false,
+            false,
+            false,
+            None,
+            &mut [(
+                onion_port,
+                format!("127.0.0.1:{}", api_port).parse().unwrap(),
+            )]
+            .iter(),
+        )
+        .await
+        .map_err(|e| {
+            Error::new(
+                ErrorKind::Other,
+                format!("failed to create onion hidden service: {}", e),
+            )
+        })?;
+
+    print_onion_service(key, onion_port);
+
+    // NOTE: Needed to keep connection with control port & hidden service running, as soon as we leave
```

Comment:
Is there no way of doing this without a loop constantly checking? Given it comes from https://github.com/teawithsand/torut/blob/master/examples/make_onion_v3.rs it may be worth opening a ticket asking.

If there is no other way, at least lets do:

```
sleep(Duration::from_secs(1)).await;
```

Reply:
Let keep it as is for now. Please do open an issue referencing this so we can tackle if we end up getting a solution, or if we figure out how to make it cleaner in the future